### ISSUE TITLE
Support writes to Azure Storage

### ISSUE THREAD

**Issue by thovoll** [delta-io/delta-rs#489]
# Description
The delta-rs Azure Storage backend currently doesn't support write operations:
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L1
As a result, Delta Tables can not be written to Azure Storage.

Several methods need to be implemented:
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L192
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L196
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L200

The delta-rs Azure Storage backend currently uses the [Blob API](https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-rest-api) to access the storage account (`ContainerClient`), which does not support atomic rename (rename_obj_noreplace). We need to use the [ADLS Gen2 API](https://docs.microsoft.com/en-us/rest/api/storageservices/data-lake-storage-gen2), which does support atomic rename. 

[Both APIs can be used in parallel](https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-multi-protocol-access) with ADLS Gen2 storage accounts, but we may want to simply switch all operations over to the ADLS Gen2 API unless there is an advantage to using the Blob API for some operations.

The other operations required by the delta-rs Azure Storage backend are:
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L122
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L141
https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/azure.rs#L157

For atomic rename to work, the actual Azure Storage Account resource being used [must have hierarchical namespaces enabled](https://docs.microsoft.com/en-us/azure/storage/blobs/create-data-lake-storage-account#enable-the-hierarchical-namespace), which is what makes it an "ADLS Gen2" storage account.

Apologies for any remaining confusion, this is a bit tricky to explain.

Also, the Azure Rust SDK [doesn't yet support write operations via the ADLS Gen API](https://github.com/Azure/azure-sdk-for-rust/issues/356), but work is underway.

**Use Case**
Write Delta Tables to Azure Storage

**Related Issue(s)**
None

**Comment by roeap** [comment]
I am considering looking into this once (if :)) #486 gets merged. MY assumption would be that `delete_obj` and `put_obj` could be implemented without the need for hierarchical namespaces enabled. However I am still a bit confused what constitutes gen2... As far as I understand, the *.dfs.core.windows.net endpoint is available on all recent (v2) storage accounts, if HNS is enabled, it becomes ADLS Gen 2, and the file_system APIs / directories become available? If HNS is a requirement, likely #60 should then be addressed as well somewhere.

**Comment by thovoll** [comment]
Thanks @roeap, that would be awesome. Your assumptions above are all correct. Only HNS supports atomic renames, which will obviate the need for [the lock service that we have to use for S3](https://github.com/delta-io/delta-rs/blob/50a53cad084f2aa65ed0d6c0f1f5ad620a785530/rust/src/storage/s3/mod.rs#L614).

**Comment by thovoll** [comment]
PR for #60 is #492

**Comment by roeap** [comment]
@thovoll, thanks for the pointers, and excited to see that the azure-sdk-for-rust repo seems to see some increased activity recently - I'm guessing pushing for a release of the crates... hopefully we can bring the azure support throughout the arrow-rust-data-ecosystem en par with S3 :).

**Comment by thovoll** [comment]
Keep track of ADLS Gen2 support in the Azure Rust SDK here: https://github.com/Azure/azure-sdk-for-rust/issues/496

**Comment by houqp** [comment]
do we know if the atomic rename in gen2 api is a cheap pointer change or will actually require data copy behind the scene? If it's the latter, we might be better off using simple put if absent semantic instead. @blogle started a good discussion around this redesign in [slack](https://delta-users.slack.com/archives/C013LCAEB98/p1628289773019100?thread_ts=1628117972.015100&cid=C013LCAEB98) for the GCS backend.

**Comment by thovoll** [comment]
Yes, _file rename_ is a cheap and atomic operation in ADLS Gen2.

**Comment by thovoll** [comment]
FYI, we now have the required operation in the Azure Rust SDK that enables us to implement `rename_obj_noreplace` in the delta-rs Azure Storage backend:
https://github.com/Azure/azure-sdk-for-rust/pull/506

I know #486 is ready to merge now, so we will be able to start working on this soon.

**Comment by andrei-ionescu** [comment]
I just want to let you know there is another 3rd party Azure SDK implementation here - https://crates.io/crates/azure_sdk_for_rust - and it has crates released.

**Comment by roeap** [comment]
I think these crates are actually the foundation for the development going on under the Azure github org right now. If you go to the repo, you'll see a notice that they migrated. Looking at the efforts in the new repo, they are working hard on stabilizing APIs and hopefully soon cutting a 0.1 release.

**Comment by thovoll** [comment]
@roeap @houqp what do you think about the auth methods we should support?

After #486 we have:
- Static auth: shared key, shared access signature (SAS), and connection string (shared key)
- Auto-refreshing auth based on DefaultAzureCredential

None of these should be used in production applications in my opinion, for various reasons. The static methods present key rotation and authz challenges. DefaultAzureCredential is just a convenience abstraction with non-deterministic behavior.

I propose we use managed identity credentials (AMI). The downside is that they are hard to use during development, but we can support `az cli` credentials for this and activate that path using a suitable switch in delta-rs, maybe an environment variable.

Making these work at runtime [requires an identity in Azure](https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-access-control-model) (which the key based methods don't), but this is the right direction since it better supports key rotation and authz.

I realize we should probably align this with how we do things for AWS and GCP, but I wanted to present this unadulterated argument first.

Any thoughts?

**Comment by houqp** [comment]
Thanks for starting this discussion @thovoll . I think we should definitely have first class support for AMI due to production requirements. @thovoll does Azure sdk support the concept of credential chain? Basically the way it works for AWS is the client will automatically identify the best auth method based on provided environment variable and credential config file across different languages and platforms: see https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html and https://github.com/rusoto/rusoto/blob/master/AWS-CREDENTIALS.md. If we have something similar provided by the azure sdk out of the box, that would be ideal. If not, we can implement a temp workaround within delta-rs to simulate something similar. For example, go for AMI if it's available, if not try other routes that are easier to prepare for development environments.

**Comment by roeap** [comment]
The DefaultCredential we use is an credential chain. Only authorization via KEY or or SAS is not part of this chain. I am not sure about priorities in the chained credential, but assume this can be either configured or achieved with a simple custom credential.

**Comment by thovoll** [comment]
Correct, [default credential chain](https://github.com/Azure/azure-sdk-for-rust/blob/7ac58a21003eb1f7c74b746c1d08dd9390d7fbe7/sdk/identity/src/token_credentials/default_credentials.rs#L134) from the Azure Rust SDK can be configured but it's still rudimentary, also in terms of the methods it supports. There's no custom chain in the Azure Rust SDK yet but the Azure .Net SDK for example has more complete [default](https://docs.microsoft.com/en-us/dotnet/api/overview/azure/identity-readme#defaultazurecredential) and [custom](https://docs.microsoft.com/en-us/dotnet/api/overview/azure/identity-readme#define-a-custom-authentication-flow-with-the-chainedtokencredential) credential chains which will be implemented in the Azure Rust SDK at some point.

Regardless, my argument is that credential chains are convenient but shouldn't be used for production code. I'd rather explicitly state what credentials I want to use than falling through a hidden switch statement based on environmental conditions. The potential 'decoy' log entries alone are distracting.

If we only support a couple of types of credentials this should be pretty easy but if we want to support all the flavors it would be a little tedious to implement a deterministic approach - but I'd probably still prefer it. Since delta-rs is a library we should probably support all the credential types (but without using a chain).

The [AWS docs](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html) that @houqp linked recommend using the default chain but also mention the other approaches:
> * Use the default credential provider chain __(recommended)__.
> * Use a specific credential provider or provider chain (or create your own).
> * Supply the credentials yourself. These can be root account credentials, IAM credentials, or temporary credentials retrieved from AWS STS.

The [Azure docs](https://docs.microsoft.com/en-us/dotnet/api/overview/azure/identity-readme#defaultazurecredential) at least hint at the tradeoff:
> Note: DefaultAzureCredential is intended to simplify getting started with the SDK by handling common scenarios with reasonable default behaviors. Developers who want more control or whose scenario isn't served by the default settings should use other credential types.

**Comment by thovoll** [comment]
As for a concrete long-term proposal (long-term because the Azure Rust SDK doesn't fully support this yet):
* Take `TokenCredential` as a parameter in `delta-rs` and pass it to the `DataLakeClient` (not possible yet but [here](https://github.com/Azure/azure-sdk-for-net/blob/c5738f34a5ded5bf4b49e380a8eec179e12a1c6b/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeServiceClient.cs#L264) is what that looks like in the Azure .Net SDK). 
* This still allows `delta-rs` users to specify a chain since `DefaultAzureCredential` is a `TokenCredential`. But it doesn't allow `delta-rs` users to use other methods like [shared keys](https://github.com/Azure/azure-sdk-for-net/blob/c5738f34a5ded5bf4b49e380a8eec179e12a1c6b/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeServiceClient.cs#L186) or [connection strings](https://github.com/Azure/azure-sdk-for-net/blob/c5738f34a5ded5bf4b49e380a8eec179e12a1c6b/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeServiceClient.cs#L152) (which the Azure SDK supports but we would be hiding from `delta-rs` users, indicating our [_industrial strength_ intent](https://docs.microsoft.com/en-us/azure/storage/common/shared-key-authorization-prevent?tabs=portal)). 
* Maybe we should support [SAS](https://github.com/Azure/azure-sdk-for-net/blob/c5738f34a5ded5bf4b49e380a8eec179e12a1c6b/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeServiceClient.cs#L225) but this method is [intended](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview) for "limited delegated access" (potentially a scenario that `delta-rs` would be used in). 
* Also need to revisit how we can auto-refresh the credential, which [will likely involve](https://github.com/Azure/azure-sdk-for-net/blob/c13662bd5365cb66a13beb9ffa32337965e8c2a7/sdk/core/Azure.Core/src/TokenCredential.cs#L22) constructing a new `DataLakeClient` at the right time.
* In `kafka-delta-ingest`, don't pass `DefaultAzureCredential` to `delta-rs` but implement a more opinionated and deterministic approach. Only support AMI (for prod) and CLI (for dev), possibly supported by a command line argument or environment variable.

Again, this is a long-term proposal but we can still discuss this in principle. If we agree, we can plot a path towards this future.

**Comment by thovoll** [comment]
As for a concrete short-term proposal:
* Obtain a token (via `DefaultAzureCredential`) in `delta-rs`
* Pass the token to `DataLakeClient` (this is the only method supported by the Azure Rust SDK at the moment)
* This is the quickest way to enable writes to Azure Storage.
* Improve from there, i.e. re-implement auto-refreshing credentials and everything else mentioned above over time.

**Comment by thovoll** [comment]
It's definitely doable and I'm sure many people do the same in Azure. Irrespective of how common it is, what do you think about the argument otherwise?

For S3, I think you meant this one?
https://github.com/delta-io/delta-rs/blob/cba4e3d316c56c65dc1cb660dc7e6d380ddd3a6b/rust/src/storage/s3/mod.rs#L425-L429

I'll take a closer look and try to keep it close.

Do we support auto-refreshing credentials in S3 and GCP?

**Comment by houqp** [comment]
Yeah, sorry @thovoll , `new_with` is the method that I was trying to use as an example. Auto refresh of credential should be supported by cloud SDK if possible.

**Comment by roeap** [comment]
The way we implemented for the Azure credentials is actually heavily inspired by the internals of the rusto s3 crate 🙂.

**Comment by thovoll** [comment]
Does the AWS and GCP SDK support auto-refresh?

**Comment by houqp** [comment]
Yes, I remember the AWS SDK has an auto refreshing credential provider builtin.

**Comment by thovoll** [comment]
I'm looking into how to best implement auto-refresh.

**Comment by thovoll** [comment]
Ok, so auto-refresh in the Azure Rust SDK will take some time, filed an issue here: https://github.com/Azure/azure-sdk-for-rust/issues/543

In the meantime, we'll have to implement auto-refresh in `delta-rs`.

**Comment by roeap** [comment]
@thovoll - this is done by now, right?
